### PR TITLE
devnet2 stack 01 Stabilize fork-choice state during block production

### PR DESCRIPTION
**Pull request by tbenr**
Stabilize fork-choice state during block production and wire late-block reorg preparation through core paths.

---

> [!NOTE]
> **Medium Risk**
> Touches validator block production and execution-layer request selection by threading a new `BlockProductionContext` (including fork-choice payload status) through core paths, which could affect block contents/production under edge conditions. Most changes are signature/record refactors with broad call-site updates and test coverage adjustments.
> 
> **Overview**
> **Stabilizes validator block production by snapshotting fork-choice inputs**: introduces `BlockProductionContext` and updates `BlockFactory`/`BlockOperationSelectorFactory` to consume a single context object (slot/state/parent root/randao/graffiti/builder boost/perf).
> 
> `ValidatorApiHandler` now captures a `ChainHead` during `prepareForBlockProduction`, uses it when fetching state for production, and carries `payloadStatus` into block building so parent execution requests are only read when the payload is known `FULL` (otherwise defaults are used). This also wires the new `ExecutionPayloadManager.getParentExecutionRequestsForBlock(..., payloadStatus)` signature through the block body builder.
> 
> Separately, `ForkChoiceState` and `PayloadBuildingAttributes` are converted to records (new accessors and `parentBeaconBlock` node), and execution client / REST event code is updated accordingly; tests are adjusted and a small `BlockProductionTestUtil` helper is added.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2c5449abc9ff07744d51f4aa6ef925c49c54ed1a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: gradle/versions.gradle [hunk 1]
```diff
@@ -37,7 +37,7 @@ dependencyManagement {
 			entry 'javalin-rendering-thymeleaf'
 		}
 
-		dependency 'io.libp2p:jvm-libp2p:1.2.2-RELEASE'
+		dependency 'io.libp2p:jvm-libp2p:develop'
```
Comment (zilm13) [inline gradle/versions.gradle]: what's our decision on this?

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java [hunk 2]
```diff
@@ -1231,6 +1211,65 @@ private boolean isLocallyCreatedBlock(final SignedBlockContainer signedBlockCont
   }
 
   private record BlockProductionPreparationContext(
```
Comment (zilm13) [inline beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java:1212]: I think it's worth to refactor both of them to outer classes

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/execution/DefaultExecutionPayloadManager.java [hunk 3]
```diff
@@ -170,10 +171,10 @@ public SafeFuture<ExecutionPayloadImportResult> importExecutionPayload(
 
   @Override
   public ExecutionRequests getParentExecutionRequestsForBlock(
-      final UInt64 slot, final Bytes32 parentRoot) {
+      final UInt64 slot, final Bytes32 parentRoot, final ForkChoicePayloadStatus payloadStatus) {
     final SpecVersion specVersion = spec.atSlot(slot);
     final UpdatableStore store = recentChainData.getStore();
-    if (!store.getForkChoiceStrategy().shouldExtendPayload(store, parentRoot)) {
+    if (!payloadStatus.equals(ForkChoicePayloadStatus.PAYLOAD_STATUS_FULL)) {
```
Comment (zilm13) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/execution/DefaultExecutionPayloadManager.java:183]: Could you add a comment with rationale here

File: storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java [hunk 4]
```diff
@@ -301,45 +297,24 @@ public SafeFuture<Optional<BeaconState>> getStateAtSlotExact(final UInt64 slot)
...
+   * advances the supplied chain-head state to the proposing slot.
+   */
+  public SafeFuture<BeaconState> getStateForBlockProduction(
+      final ChainHead chainHead, final UInt64 slot) {
+    return chainHead
+        .getState()
+        .thenApply(
+            state -> {
+              try {
+                return spec.processSlots(state, slot);
```
Comment (zilm13) [inline storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java]: why we are not using `getStateAtSlotExact` anymore? it has some caching i think

File: storage/src/main/java/tech/pegasys/teku/storage/protoarray/ForkChoiceStrategy.java [hunk 5]
```diff
@@ -257,14 +258,22 @@ && getForkChoiceModel(protoNode.getBlockSlot()).isHeadCandidate(protoNode)
...
+      final Optional<ChainHead> proposingOnHead,
       final UInt64 currentEpoch,
       final Checkpoint justifiedCheckpoint,
       final Checkpoint finalizedCheckpoint) {
     protoArrayLock.readLock().lock();
     try {
       final ProtoNode headNode =
-          protoArray.findOptimisticHead(
-              currentEpoch, justifiedCheckpoint, finalizedCheckpoint, headSelectionContext);
+          proposingOnHead
```
Comment (zilm13) [inline storage/src/main/java/tech/pegasys/teku/storage/protoarray/ForkChoiceStrategy.java:268]: maybe add a test on this?

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifierImpl.java [hunk 6]
```diff
@@ -159,12 +163,19 @@ private SafeFuture<Optional<ExecutionPayloadContext>> internalGetPayloadId(
...
       return forkChoiceUpdateData.getExecutionPayloadContext();
     } else if (parentExecutionHash.isZero() && !forkChoiceUpdateData.hasTerminalBlockHash()) {
       // Pre-merge so ok to use default payload
       return SafeFuture.completedFuture(Optional.empty());
     } else {
+      // TODO: with the pinned forkChoiceUpdateData there is no point in trying to refresh.
+      // Now we should have a stable forkChoiceUpdateData which must be compatible with
+      // parentBeaconBlockRoot and slot by blockSlot. So we can just throw here.
+
       // Request a new payload with refreshed forkChoiceState and payloadBuildingAttributes
```
Comment (zilm13) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifierImpl.java:179]: maybe better detach it to the issue?

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java [hunk 7]
```diff
@@ -1082,7 +1104,103 @@ SafeFuture<Void> prepareForBlockProduction(
...
+            // Run preparation in parallel; await it before returning so block-body construction
+            // sees the updated operation pools.
+            preparation =
+                lateBlockReorgPreparationHandler
+                    .onLateBlockReorgPreparation(proposerHead.getSlot(), canonicalHead.getRoot())
+                    .thenPeek(__ -> performance.lateBlockReorgPreparationCompleted());
+          } else {
+            preparation = SafeFuture.COMPLETE;
+          }
+          notifyForkChoiceUpdatedAndOptimisticSyncingChanged(
```
Comment (zilm13) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java:1146]: should we if it's not changed?

File: storage/src/test/java/tech/pegasys/teku/storage/client/CombinedChainDataClientTest.java [hunk 8]
```diff
@@ -157,126 +150,16 @@ void getsEarliestAvailableBlobSidecarSlot() {
...
-
-    // should be pending until late block reorg preparation completes
-    assertThat(future).isNotDone();
-    verify(onLateBlockReorgPreparationCompleted, never()).run();
-    // should retrieve state while waiting
-    verify(store).retrieveBlockState(proposerHead);
-
-    lateBlockReorgPreparationFuture.complete(null);
-    verify(onLateBlockReorgPreparationCompleted).run();
+  void getStateForBlockProduction_processesChainHeadState() {
```
Comment (zilm13) [inline storage/src/test/java/tech/pegasys/teku/storage/client/CombinedChainDataClientTest.java]: shouldn't we move late block reorg tests to `ForkChoiceTest`?

File: gradle/versions.gradle (hunk 1 above)
Comment (StefanBratanov) [reply gradle/versions.gradle]: Never do this, makes our build non-deterministic, we should revert this

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java (hunk 2 above)
Comment (tbenr) [reply beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java:1212]: i think the preparation context is really just used internally in the handler, but I agree on the other one...

File: storage/src/test/java/tech/pegasys/teku/storage/client/CombinedChainDataClientTest.java (hunk 8 above)
Comment (tbenr) [reply storage/src/test/java/tech/pegasys/teku/storage/client/CombinedChainDataClientTest.java]: I agree, i did it before when i was dealing with forkchoice gloas changes, but then I decided to postpone it to reduce noise. We can do it once things settle. I can open an issue

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/execution/DefaultExecutionPayloadManager.java (hunk 3 above)
Comment (tbenr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/execution/DefaultExecutionPayloadManager.java:183]: sure. i'll do

File: gradle/versions.gradle (hunk 1 above)
Comment (tbenr) [reply gradle/versions.gradle]: Ye I can revert it. there was a method name typo that was annoying me (and we can't point to master on devnets until we do the release)

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/execution/DefaultExecutionPayloadManager.java (hunk 3 above)
Comment (tbenr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/execution/DefaultExecutionPayloadManager.java:183]: but to be fair it is pretty obvious

File: storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java (hunk 4 above)
Comment (tbenr) [reply storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java]: the idea is that we get the state from `ChainHead` directly.

I believe there are minimal differences, both end up calling `getOrRegenerateBlockAndState`

image

I'll doublecheck carefully

Comment (tbenr) [reply storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java]: It turns out that this was causing slot advancing (processSlot call) happening twice, which is bad in epoch transition especially. Fixing.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java (hunk 7 above)
Comment (tbenr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java:1146]: no, here we want to always call because it notifies the FCUNotifier that we started block production. I'll add a comment.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifierImpl.java (hunk 6 above)
Comment (tbenr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifierImpl.java:179]: opened an issue

**Review by zilm13 — DISMISSED** [review]
LGTM, just 2 nits:
- https://github.com/Consensys/teku/pull/10683#discussion_r3209166931 makes sense, we could remove it as it's already checked when context was created
- please, create an issue, we need more tests for all these changes

**Review by zilm13 — APPROVED** [review]
Full throttle!
